### ISSUE TITLE
AWS ENI: Pod assigned IP that is not attached to an ENI

### ISSUE THREAD

**Issue by John-Moya** [cilium/cilium#13412]
Two pods on my cluster are geting an IP that is not attached to an ENI

Cilium version is 1.8.4
EKS Kubernetes version 1.16

Observed this issue after upgrading from 1.8.3 to 1.8.4 and adding nodes to the cluster

This output shows a CiliumNode where the rbac-manager pod is assigned the IP  10.80.17.39 and that IP is not associated with any of the active ENIs 

```
apiVersion: cilium.io/v2
kind: CiliumNode
metadata:
  creationTimestamp: "2020-10-05T15:17:10Z"
  generation: 19
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: m5.large
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-east-1
    failure-domain.beta.kubernetes.io/zone: us-east-1a
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: ip-10-80-17-191.us-east-1.compute.internal
    kubernetes.io/os: linux
    nite.tradestation.io/role: worker
  name: ip-10-80-17-191.ec2.internal
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: ip-10-80-17-191.ec2.internal
    uid: cf85a4d5-8d93-4ba4-88d5-9ac8251df556
  resourceVersion: "7640153"
  selfLink: /apis/cilium.io/v2/ciliumnodes/ip-10-80-17-191.ec2.internal
  uid: 1b6030a4-eb02-4634-8410-a3e1f9051103
spec:
  addresses:
  - ip: 10.80.17.191
    type: InternalIP
  - ip: 10.80.17.232
    type: CiliumInternalIP
  azure: {}
  encryption: {}
  eni:
    availability-zone: us-east-1a
    delete-on-termination: true
    first-interface-index: 1
    instance-type: m5.large
    subnet-tags:
      NetworkType: private
    vpc-id: vpc-0e030555a019aefa4
  health:
    ipv4: 10.80.17.16
  instance-id: i-0466c2a16aaa4eb42
  ipam:
    podCIDRs:
    - 10.191.0.0/16
    pool:
      10.80.17.16:
        resource: eni-019ecf4a65359906b
      10.80.17.24:
        resource: eni-019ecf4a65359906b
      10.80.17.28:
        resource: eni-019ecf4a65359906b
      10.80.17.30:
        resource: eni-0d938326ff46f5e20
      10.80.17.85:
        resource: eni-0d938326ff46f5e20
      10.80.17.86:
        resource: eni-0d938326ff46f5e20
      10.80.17.100:
        resource: eni-0d938326ff46f5e20
      10.80.17.102:
        resource: eni-019ecf4a65359906b
      10.80.17.108:
        resource: eni-0d938326ff46f5e20
      10.80.17.112:
        resource: eni-019ecf4a65359906b
      10.80.17.116:
        resource: eni-0d938326ff46f5e20
      10.80.17.136:
        resource: eni-019ecf4a65359906b
      10.80.17.186:
        resource: eni-019ecf4a65359906b
      10.80.17.207:
        resource: eni-019ecf4a65359906b
      10.80.17.232:
        resource: eni-0d938326ff46f5e20
      10.80.17.234:
        resource: eni-019ecf4a65359906b
      10.80.17.239:
        resource: eni-0d938326ff46f5e20
      10.80.17.251:
        resource: eni-0d938326ff46f5e20
      10.80.17.253:
        resource: eni-019ecf4a65359906b
    pre-allocate: 8
status:
  azure: {}
  eni:
    enis:
      eni-019ecf4a65359906b:
        addresses:
        - 10.80.17.16
        - 10.80.17.253
        - 10.80.17.136
        - 10.80.17.234
        - 10.80.17.186
        - 10.80.17.28
        - 10.80.17.24
        - 10.80.17.207
        - 10.80.17.102
        - 10.80.17.112
        description: Cilium-CNI (i-0466c2a16aaa4eb42)
        id: eni-019ecf4a65359906b
        ip: 10.80.17.16
        mac: 02:30:e9:1e:a2:79
        number: 1
        security-groups:
        - sg-0a521e746d55bca88
        subnet:
          cidr: 10.80.17.0/24
          id: subnet-0ec1716995dd5c73b
        vpc:
          id: vpc-0e030555a019aefa4
          primary-cidr: 10.80.16.0/22
      eni-0cd1b87a2f1c93097:
        addresses:
        - 10.80.17.191
        id: eni-0cd1b87a2f1c93097
        ip: 10.80.17.191
        mac: 02:1b:da:9c:0e:51
        security-groups:
        - sg-0a521e746d55bca88
        subnet:
          cidr: 10.80.17.0/24
          id: subnet-0ec1716995dd5c73b
        vpc:
          id: vpc-0e030555a019aefa4
          primary-cidr: 10.80.16.0/22
      eni-0d938326ff46f5e20:
        addresses:
        - 10.80.17.116
        - 10.80.17.232
        - 10.80.17.251
        - 10.80.17.239
        - 10.80.17.100
        - 10.80.17.86
        - 10.80.17.108
        - 10.80.17.30
        - 10.80.17.85
        description: Cilium-CNI (i-0466c2a16aaa4eb42)
        id: eni-0d938326ff46f5e20
        ip: 10.80.17.116
        mac: 02:9f:02:c6:08:8d
        number: 2
        security-groups:
        - sg-0a521e746d55bca88
        subnet:
          cidr: 10.80.17.0/24
          id: subnet-0ec1716995dd5c73b
        vpc:
          id: vpc-0e030555a019aefa4
          primary-cidr: 10.80.16.0/22
  ipam:
    operator-status: {}
    used:
      10.80.17.16:
        owner: health
        resource: eni-019ecf4a65359906b
      10.80.17.28:
        owner: kube-system/efs-provisioner-gp-5b488cf9bf-pwcvq
        resource: eni-019ecf4a65359906b
      10.80.17.39:
        owner: kube-system/rbac-manager-5f56b4d9c8-lkvxw
        resource: eni-0d938326ff46f5e20
      10.80.17.100:
        owner: kube-system/hubble-relay-56dcd9c4b-pmcrm
        resource: eni-0d938326ff46f5e20
      10.80.17.116:
        owner: kube-system/fluentd-xmtsc
        resource: eni-0d938326ff46f5e20
      10.80.17.136:
        owner: default/pod-to-a-denied-cnp-6967cb6f7f-qt4zx
        resource: eni-019ecf4a65359906b
      10.80.17.232:
        owner: router
        resource: eni-0d938326ff46f5e20
      10.80.17.234:
        owner: nite-opa/gatekeeper-controller-manager-85c6977484-4vjkv
        resource: eni-019ecf4a65359906b
      10.80.17.239:
        owner: kube-system/prometheus-prometheus-prometheus-oper-prometheus-0
        resource: eni-0d938326ff46f5e20
      10.80.17.251:
        owner: kube-system/datadog-c82cg
        resource: eni-0d938326ff46f5e20
      10.80.17.253:
        owner: default/pod-to-b-multi-node-clusterip-6fcdb496c7-vsn68
        resource: eni-019ecf4a65359906b
```

**Comment by jaffcheng** [comment]
Hi @John-Moya ! We are facing the same issue and in our case, it's due to cilium-operator releases and unassigns the pod IP at the same time the pod IP is allocated to a new pod. Have you enabled the operator flag `aws-release-excess-ips`?

**Comment by John-Moya** [comment]
@jaffcheng I will test and post an update as soon as possible 👍

**Comment by John-Moya** [comment]
The issues is still present using the recommended flag on the operator `aws-release-excess-ips`, our configuration is the following

- EKS cluster
- Instance Type m5.xlarge
- Kubernetes Version 1.16
- Cilium 1.8.3
- CRD backed IPAM 

Reproduction Steps

1. Deploy this file into the cluster
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
  namespace: default  
spec:
  replicas: 10
  selector:
    matchLabels:
      app: nginx
  template:
    
    metadata:
      labels:
        app: nginx
    spec:
      nodeName: <one-worker-node>
      containers:
      - name: nginx
        image: nginx:1.7.9
        ports:
        - containerPort: 80

```
2. Then run this script and keep it running

```
import subprocess
from time import sleep

def killpods():
    while True:
        subprocess.run(['kubectl','scale','deployment','nginx-deployment','--replicas=10',])
        sleep(5)
        subprocess.run(['kubectl','scale','deployment','nginx-deployment','--replicas=20',])

killpods()
```
3. Grep the cilium operator pods for `Unable to maintain ip pool of node` (we use `watch`


 with grep)

The following info will appear

```
level=warning msg="Unable to maintain ip pool of node" error="InvalidParameterValue: Some of the specified addresses are not assigned to interface eni-08979abb2c28fb6a4\n\tstatus code: 400, request id: bc588ad1-a579-
482d-99dd-d344dd860455" instanceID=i-0257df4fe99f48a58 name=ip-10-80-17-165.ec2.internal subsys=ipam
level=warning msg="Unable to maintain ip pool of node" error="InvalidParameterValue: Some of the specified addresses are not assigned to interface eni-05e2b76b341d9d269\n\tstatus code: 400, request id: d5d79dde-b17c-
4e2e-b5aa-3e89f11c10f9" instanceID=i-0b41659593d672463 name=ip-10-80-18-191.ec2.internal subsys=ipam
level=warning msg="Unable to maintain ip pool of node" error="InvalidParameterValue: Some of the specified addresses are not assigned to interface eni-04d933fee6c55e877\n\tstatus code: 400, request id: 41d34746-cf6b-
469a-b1a3-da9f1e45b432" instanceID=i-0257df4fe99f48a58 name=ip-10-80-17-165.ec2.internal subsys=ipam
level=warning msg="Unable to maintain ip pool of node" error="InvalidParameterValue: Some of the specified addresses are not assigned to interface eni-0616f6a4f54d140d9\n\tstatus code: 400, request id: 4d3852c5-6a42-
4f98-995e-2f1b0cdbbc67" instanceID=i-021ad793cf9981274 name=ip-10-80-18-254.ec2.internal subsys=ipam
level=warning msg="Unable to maintain ip pool of node" error="InvalidParameterValue: Some of the specified addresses are not assigned to interface eni-08979abb2c28fb6a4\n\tstatus code: 400, request id: e251073a-6a8e-
4344-b83e-5b49c08e4ca4" instanceID=i-0257df4fe99f48a58 name=ip-10-80-17-165.ec2.internal subsys=ipam
```

**Comment by jaffcheng** [comment]
Thanks for testing! If it happens too frequently, maybe disabling `aws-release-excess-ips` could be a degradation. I'll try to dig deeper.

**Comment by jaffcheng** [comment]
It seems the issue is that cilium-agent and cilium-operator may be out of sync in the following order:

1. cilium-agent allocates `ip_address1` from node IP pool and triggers sync upstream, pod becomes ready and reachable
https://github.com/cilium/cilium/blob/2ec73052ecbe9b24e7a03923e501fa20c9a1e3a4/pkg/ipam/crd.go#L540
2. the trigger doesn't call sync upstream immediately and there is a custom resource update rate defined as `15 seconds`
https://github.com/cilium/cilium/blob/2ec73052ecbe9b24e7a03923e501fa20c9a1e3a4/pkg/ipam/crd.go#L56
3. at this point, cilium-operator checks excess IP depending on used IPs (**which is out of sync**) and available IPs from ciliumnode custom resource. If operator finds IP excess, there is a chance `ip_address1` gets released and pod becomes unreachable but ready.
https://github.com/cilium/cilium/blob/2ec73052ecbe9b24e7a03923e501fa20c9a1e3a4/pkg/ipam/node_manager.go#L200-L217

**Comment by christarazi** [comment]
Thanks for the write-up @jaffcheng. That definitely seems plausible. One thing that could be interesting is to run the agent with race detection enabled to see if this level of churn surfaces any problems.

**Comment by christarazi** [comment]
@John-Moya Are you still experiencing this with Cilium v1.9.5?

**Comment by John-Moya** [comment]
@christarazi using the aws-release-excess-ips flag since v1.9.4 and we have not observed the issue

**Comment by dctrwatson** [comment]
> @John-Moya Are you still experiencing this with Cilium v1.9.5?

Have been having issues with this in 1.9.5 and 1.9.6 so far

**Comment by christarazi** [comment]
@dctrwatson Do you have reproduction steps and a sysdump?

**Comment by dctrwatson** [comment]
> @dctrwatson Do you have reproduction steps and a sysdump?

repro is to restart 1000+ pods at the same time (the higher concurrency, the higher the likelihood)

At least a few hundred pods will get an IP that has been released by the operator.

Deleting the bad pods so they are rescheduled, fixes them. Well, as long as too many aren't deleted at the same time.

No sysdump, just that in the last batch of broken pods, I was able to determine they all had IPs that were no longer assigned to an ENI.
